pkg/controller/elasticsearch/remotecluster/elasticsearch.go

```diff
@@ -25,88 +26,143 @@ var log = logf.Log.WithName("remotecluster")
 const enterpriseFeaturesDisabledMsg = "Remote cluster is an enterprise feature. Enterprise features are disabled"
 
 // UpdateSettings updates the remote clusters in the persistent settings by calling the Elasticsearch API.
+// See the documentation of updateSettingsInternal for more information about the algorithm.
 func UpdateSettings(
 	ctx context.Context,
 	c k8s.Client,
 	esClient esclient.Client,
 	eventRecorder record.EventRecorder,
 	licenseChecker license.Checker,
 	es esv1.Elasticsearch,
-) error {
+) (bool, error) {
 	span, _ := apm.StartSpan(ctx, "update_remote_clusters", tracing.SpanTypeApp)
 	defer span.End()
 
-	expectedRemoteClusters := getExpectedRemoteClusters(es)
+	remoteClustersInSpec := getRemoteClustersInSpec(es)
 	enabled, err := licenseChecker.EnterpriseFeaturesEnabled()
 	if err != nil {
-		return err
+		return true, err
 	}
-	if !enabled && len(expectedRemoteClusters) > 0 {
+	if !enabled && len(remoteClustersInSpec) > 0 {
 		log.Info(
 			enterpriseFeaturesDisabledMsg,
 			"namespace", es.Namespace, "es_name", es.Name,
 		)
 		eventRecorder.Eventf(&es, corev1.EventTypeWarning, events.EventAssociationError, enterpriseFeaturesDisabledMsg)
-		return nil
+		return false, nil
 	}
 
-	currentRemoteClusters, err := getCurrentRemoteClusters(es)
+	return updateSettingsInternal(remoteClustersInSpec, c, esClient, es)
+}
+
+// updateSettingsInternal updates remote clusters in Elasticsearch. It also keeps track of any remote clusters which
+// have been declared in the Elasticsearch spec. The purpose is to delete remote clusters which were managed by
+// the operator but are not desired anymore, without removing the ones which have been added by the user.
+// The following algorithm is used:
+// 1. Get the list of the previously declared remote clusters from the annotation
+// 2. Ensure that all remote clusters in the Elasticsearch spec are present in the annotation
+// 3. For each remote cluster in the annotation which is not the Spec, either:
+//   3.1 Schedule its deletion from the Elasticsearch settings
+//   3.2 Otherwise remove it from the annotation
+// 4. Update the annotation on the Elasticsearch object
+// 5. Apply the settings through the Elasticsearch API
+func updateSettingsInternal(
+	remoteClustersInSpec map[string]esv1.RemoteCluster,
+	c k8s.Client,
+	esClient esclient.Client,
+	es esv1.Elasticsearch,
+) (bool, error) {
```

Comment:
I would also appreciate a comment on that bool here (or maybe just use a named result eg. `(requeue bool, err error)`)